src/NuGet.Clients/VisualStudio.Facade/NuGetUIThreadHelper.cs

```diff
@@ -3,22 +3,54 @@
 
 using System;
 using System.Threading;
-using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.ProjectSystem;
 using Microsoft.VisualStudio.Threading;
 
 namespace NuGet.PackageManagement.UI
 {
     public static class NuGetUIThreadHelper
     {
-        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope")]
-        public static void SetCustomJoinableTaskFactory(Thread mainThread, SynchronizationContext synchronizationContext)
+        private static JoinableTaskFactory _joinableTaskFactory;
+
+        /// <summary>
+        /// Returns the static instance of JoinableTaskFactory set by SetJoinableTaskFactoryFromService.
+        /// </summary>
+        public static JoinableTaskFactory JoinableTaskFactory
         {
-            // If ThreadHelper.JoinableTaskFactory is not null, do not allow a custom JoinableTaskFactory to be set here
-            if (GetThreadHelperJoinableTaskFactorySafe() != null)
+            get
             {
-                throw new ArgumentException(Resources.CustomJoinableTaskFactoryNotAllowed);
+                return _joinableTaskFactory;
             }
+        }
 
+        /// <summary>
+        /// Retrieve the CPS enabled JoinableTaskFactory for the current version of Visual Studio.
+        /// This overrides the default VsTaskLibraryHelper.ServiceInstance JTF.
+        /// </summary>
+        public static void SetJoinableTaskFactoryFromService(IServiceProvider serviceProvider)
+        {
+            var componentModel = serviceProvider.GetService(typeof(SComponentModel)) as IComponentModel;
+            var projectServiceAccessor = componentModel.GetService<IProjectServiceAccessor>();
+
+#if VS14
+            // Use IThreadHandling.AsyncPump for Visual Studio 2015
+            ProjectService projectService = projectServiceAccessor.GetProjectService();
+            IThreadHandling threadHandling = projectService.Services.ThreadingPolicy;
+            _joinableTaskFactory = service.AsyncPump;
+#else
+            // Use IProjectService for Visual Studio 2017
+            IProjectService projectService = projectServiceAccessor.GetProjectService();
```

Comment:
That API isn't currently in our dev14 build. Also, the threading docs say to use this for dev15.